### PR TITLE
Issue/624

### PR DESCRIPTION
**Pull request by fjarrett**
Implements #624 

This PR adds a custom "Restriction" setting to Ninja Forms to enable/disable referral tracking.

screen_shot_2015-07-22_at_11_29_51_am

screen shot 2015-07-22 at 11 24 35 am

Just as we did in #609 (to maintain backwards compatibility) the default value for this setting is **checked**.

As such, I have also provided a DB upgrade script to set this setting for each form if it does not exist.

As an aside, there were also a lot of incorrect docblocks in the Ninja Forms integration class, so I fixed those up as well.

### CODE REVIEW

**Comment by fjarrett** [comment]
@pippinsplugins Ready for code review.

File: includes/integrations/class-ninja-forms.php [hunk 1]
```diff
@@ -16,40 +17,47 @@ public function init() {
...
 	public function add_referral( $sub_id ) {
 
 		global $ninja_forms_processing;
 
-		if( $this->was_referred() ) {
-
-			if ( $this->is_affiliate_email( $this->get_submitted_email() ) ) {
-				return; // Customers cannot refer themselves
-			}
+		if (
```
Comment (pippinsplugins) [inline includes/integrations/class-ninja-forms.php]: Can we break this up into multiple return statements please? Makes it much easier to read.

Comment (fjarrett) [inline includes/integrations/class-ninja-forms.php]: Done in https://github.com/affiliatewp/AffiliateWP/commit/d46ff4a7b9f23c15f0926202b661379bc66594a1

**Comment by fjarrett** [comment]
@pippinsplugins Just to explain my thought process a little...I honestly don't think this setting really fits under "Display" or "Restrictions" but since Ninja Forms does not provide any hooks to add custom setting metabox sections, I had to pick one. Since this setting is a "allow" type setting, I decided on the Restrictions section.

**Comment by pippinsplugins** [comment]
Restrictions is fine.

File: includes/integrations/class-ninja-forms.php [hunk 2]
```diff
@@ -16,40 +17,50 @@ public function init() {
...
-				return; // Customers cannot refer themselves
-			}
+		if ( ! $ninja_forms_processing->get_form_setting( 'affwp_allow_referrals' ) ) {
+			return;
+		}
 
-			$description    = $ninja_forms_processing->get_form_setting( 'form_title' );
-			$total          = $this->get_total();
-			$referral_total = $this->calculate_referral_amount( $total, $sub_id, $ninja_forms_processing->get_form_ID() );
+		if ( ! $this->was_referred() ) {
```
Comment (pippinsplugins) [inline includes/integrations/class-ninja-forms.php]: Let's check if was referred first. It's a faster check. No need to perform a DB query if unnecessary.

Comment (fjarrett) [inline includes/integrations/class-ninja-forms.php]: Good call. I wondered the exact same thing. Done in https://github.com/affiliatewp/AffiliateWP/commit/1968470

**Comment by pippinsplugins** [comment]
Looks good.

@sumobi can you test this further?

**Comment by sumobi** [comment]
@pippinsplugins @fjarrett Works great for me. There's just one small issue I can see. We need to check for Ninja Forms or we'll get this:

> 23-Jul-2015 01:20:01 UTC] WordPress database error Table 'affwp.wp_46_nf_objects' doesn't exist for query SELECT id FROM wp_46_nf_objects WHERE type = 'form'; made by require_once('wp-admin/admin.php'), do_action('admin_init'), call_user_func_array, Affiliate_WP_Upgrades->init, Affiliate_WP_Upgrades->v17_upgrades, Affiliate_WP_Upgrades->v17_upgrade_nforms

**Comment by sumobi** [comment]
As per https://github.com/affiliatewp/AffiliateWP/issues/609#issuecomment-103664942 I still feel this is something that should be turned on when creating a form, not enabled by default. For backwards compatibility it should be enabled for all existing forms. Does anyone agree? :)

**Comment by fjarrett** [comment]
I don't have a strong opinion either way, but I tend to agree more with what you're describing @sumobi.

By turning it on by default we are making an assumption that the _majority_ of forms created will be for ordering, which seems unlikely to me. In actuality if we really wanted to make the default "enabled" then the checkbox should be for explicitly "disabling" it.

**Comment by sumobi** [comment]
@fjarrett My thoughts exactly

**Comment by pippinsplugins** [comment]
Seconded

**Comment by fjarrett** [comment]
@sumobi Oh wow, thanks for the testing on multisite. It appears you can't "network activate" NF on multisite at all _and_ when you single-site activate it they are not creating blog-specific tables! Very interesting.

**Comment by fjarrett** [comment]
@sumobi OK this PR should be good to go now for your multisite setup.

I also applied the same defaults we discussed in this thread to #609 issue via a new #792 PR.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by sumobi** [comment]
@fjarrett I'm still seeing the notice above (https://github.com/affiliatewp/AffiliateWP/pull/790#issuecomment-123926657). I have NF installed on a sub-site, AffiliateWP is network activated.

**Comment by fjarrett** [comment]
@sumobi Hmmm interesting. Can you check your DB for me and see if the NF tables have blog prefixes on them or not? e.g. `wp_42_nf_objects`

**Comment by sumobi** [comment]
Prefixes, here's my Ninja Forms sub-site that I test on:

screen shot 2015-07-31 at 10 44 52 am

**Comment by fjarrett** [comment]
@sumobi Thanks!

**Comment by fjarrett** [comment]
OK @sumobi I've changed the upgrade script in https://github.com/affiliatewp/AffiliateWP/commit/fc42861de838ce4e59f8629d85012c22580c4b8a so it will now respect blog-specific tables. Please test again.

**Comment by pippinsplugins** [comment]
:+1L

**Comment by sumobi** [comment]
@fjarrett perfect, notice has gone away

**Pull request merged**